File: packages/oxlint-plugin-react-doctor/src/plugin/rules/design/no-repeated-container-text.ts
import {
  REPEATED_CONTAINER_TEXT_MAX_DESCENDANT_COUNT,
  REPEATED_CONTAINER_TEXT_MAX_LENGTH,
  REPEATED_CONTAINER_TEXT_MIN_COUNT,
  REPEATED_CONTAINER_TEXT_MIN_LENGTH,
} from "../../constants/design.js";
import { TAILWIND_NAMED_BREAKPOINTS } from "../../constants/tailwind.js";
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";
import { getJsxAttributeName } from "../../utils/get-jsx-attribute-name.js";
import { getStringLiteralAttributeValue } from "../../utils/get-string-literal-attribute-value.js";
import { getTailwindVisibilityAtBreakpoints } from "../../utils/get-tailwind-visibility-at-breakpoints.js";
import { getTailwindVisibilityEffect } from "../../utils/get-tailwind-visibility-effect.js";
import { hasJsxSpreadAttribute } from "../../utils/has-jsx-spread-attribute.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { hasCapabilityOrUnspecified } from "../../utils/get-react-doctor-setting.js";
import { parseTailwindClassNameToken } from "../../utils/parse-tailwind-class-name-token.js";
import type { RuleContext } from "../../utils/rule-context.js";
import { splitTailwindClassName } from "../../utils/split-tailwind-class-name.js";
import { getEffectiveStyleProperty } from "./utils/get-effective-style-property.js";
import { getEffectiveTailwindClassNameToken } from "./utils/get-effective-tailwind-class-name-token.js";
import { getInlineStyleExpression } from "./utils/get-inline-style-expression.js";
import { getStringFromClassNameAttr } from "./utils/get-string-from-class-name-attr.js";
import { getStylePropertyStringValue } from "./utils/get-style-property-string-value.js";
import { isTailwindCardSurface } from "./utils/is-tailwind-card-surface.js";

const REPEATED_TEXT_CONTAINER_NAMES = new Set(["article", "aside", "div", "section"]);
const REPEATED_TEXT_SKIPPED_NAMES = new Set([
  "a",
  "button",
  "canvas",
  "code",
  "datalist",
  "dd",
  "dl",
  "dt",
  "figure",
  "input",
  "kbd",
  "label",
  "menu",
  "nav",
  "ol",
  "option",
  "optgroup",
  "pre",
  "samp",
  "select",
  "summary",
  "svg",
  "table",
  "tbody",
  "td",
  "textarea",
  "tfoot",
  "th",
  "thead",
  "tr",
  "ul",
]);
const REPEATED_TEXT_SKIPPED_ROLES = new Set([
  "button",
  "cell",
  "grid",
  "gridcell",
  "graphics-document",
  "graphics-symbol",
  "img",
  "link",
  "list",
  "listbox",
  "listitem",
  "menu",
  "menubar",
  "navigation",
  "progressbar",
  "radiogroup",
  "row",
  "rowgroup",
  "table",
  "tablist",
  "tree",
  "treeitem",
  "diagram",
]);
const CONTENT_ATTRIBUTE_NAMES = new Set(["children", "dangerouslySetInnerHTML"]);
const DATA_VISUALIZATION_CLASS_PATTERN =
  /(?:^|[-_\s])(?:chart|graph|heatmap|plot|visualization)(?:$|[-_\s])/i;
const VISUALLY_HIDDEN_CLASS_NAMES = new Set(["screen-reader-only", "sr-only", "visually-hidden"]);
const VISUALLY_HIDDEN_UTILITY_NAMES = new Set(["not-sr-only", "sr-only"]);
const TAILWIND_NAMED_BREAKPOINT_SET = new Set(TAILWIND_NAMED_BREAKPOINTS);

interface RepeatedTextOccurrence {
  node: EsTreeNode;
  signature: string;
}

interface RepeatedTextCollection {
  descendantCount: number;
  isStatic: boolean;
  occurrencesByText: Map<string, RepeatedTextOccurrence[]>;
}

const getNativeElementName = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
): string | null => {
  if (!isNodeOfType(openingElement.name, "JSXIdentifier")) return null;
  const elementName = openingElement.name.name;
  return elementName === elementName.toLowerCase() ? elementName : null;
};

const hasContentAttribute = (openingElement: EsTreeNodeOfType<"JSXOpeningElement">): boolean =>
  openingElement.attributes.some(
    (attribute) =>
      isNodeOfType(attribute, "JSXAttribute") &&
      CONTENT_ATTRIBUTE_NAMES.has(getJsxAttributeName(attribute.name) ?? ""),
  );

const isStaticallyHidden = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
  classNameValue: string | null,
  context: RuleContext,
): boolean => {
  const hiddenAttribute = getAuthoritativeJsxAttribute(openingElement.attributes, "hidden");
  if (hiddenAttribute) {
    if (!hiddenAttribute.value) return true;
    const hiddenValue = getStringLiteralAttributeValue(hiddenAttribute);
    if (hiddenValue === null || hiddenValue.toLowerCase() !== "false") return true;
  }
  const ariaHiddenAttribute = getAuthoritativeJsxAttribute(
    openingElement.attributes,
    "aria-hidden",
  );
  if (ariaHiddenAttribute) {
    const ariaHiddenValue = getStringLiteralAttributeValue(ariaHiddenAttribute);
    if (ariaHiddenValue === null || ariaHiddenValue.toLowerCase() === "true") return true;
  }
  const styleAttribute = getAuthoritativeJsxAttribute(openingElement.attributes, "style");
  if (styleAttribute) {
    const styleExpression = getInlineStyleExpression(styleAttribute, context.scopes);
    if (!styleExpression) return false;
    const displayProperty = getEffectiveStyleProperty(styleExpression.properties, "display");
    const visibilityProperty = getEffectiveStyleProperty(styleExpression.properties, "visibility");
    if (
      (displayProperty &&
        getStylePropertyStringValue(displayProperty)?.trim().toLowerCase() === "none") ||
      (visibilityProperty &&
        ["collapse", "hidden"].includes(
          getStylePropertyStringValue(visibilityProperty)?.trim().toLowerCase() ?? "",
        ))
    ) {
      return true;
    }
    if (
      (displayProperty && getStylePropertyStringValue(displayProperty) === null) ||
      (visibilityProperty && getStylePropertyStringValue(visibilityProperty) === null)
    ) {
      return true;
    }
  }
  if (!classNameValue) return false;
  const classNameTokens = splitTailwindClassName(classNameValue);
  if (classNameTokens.some((classNameToken) => VISUALLY_HIDDEN_CLASS_NAMES.has(classNameToken))) {
    return true;
  }
  const effectiveScreenReaderUtility = getEffectiveTailwindClassNameToken(
    classNameTokens,
    (utility) => VISUALLY_HIDDEN_UTILITY_NAMES.has(utility),
  );
  if (effectiveScreenReaderUtility === "sr-only") return true;
  const visibilityAtBreakpoints = getTailwindVisibilityAtBreakpoints(classNameValue);
  return visibilityAtBreakpoints?.every((isVisible) => !isVisible) ?? false;
};

const hasResponsiveVisibility = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
): boolean => {
  const classNameValue = getStringFromClassNameAttr(openingElement);
  if (!classNameValue) return false;
  return splitTailwindClassName(classNameValue)
    .map(parseTailwindClassNameToken)
    .some(({ utility, variants }) => {
      const isVisibilityUtility =
        getTailwindVisibilityEffect(utility).status !== "not-relevant" ||
        VISUALLY_HIDDEN_UTILITY_NAMES.has(utility);
      if (!isVisibilityUtility) return false;
      return variants.some(
        (variant) =>
          TAILWIND_NAMED_BREAKPOINT_SET.has(variant) ||
          (variant.startsWith("max-") &&
            TAILWIND_NAMED_BREAKPOINT_SET.has(variant.slice("max-".length))) ||
          /^(?:min|max)-\[.+\]$/.test(variant),
      );
    });
};

const shouldSkipElement = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
  isRoot: boolean,
  context: RuleContext,
): boolean => {
  const elementName = getNativeElementName(openingElement);
  if (!elementName || REPEATED_TEXT_SKIPPED_NAMES.has(elementName)) return true;
  if (hasJsxSpreadAttribute(openingElement.attributes) || hasContentAttribute(openingElement)) {
    return true;
  }
  const classNameAttribute = getAuthoritativeJsxAttribute(openingElement.attributes, "className");
  const classNameValue = getStringFromClassNameAttr(openingElement);
  if (classNameAttribute && classNameValue === null) return true;
  if (
    (classNameValue && DATA_VISUALIZATION_CLASS_PATTERN.test(classNameValue)) ||
    isStaticallyHidden(openingElement, classNameValue, context)
  ) {
    return true;
  }
  const roleAttribute = getAuthoritativeJsxAttribute(openingElement.attributes, "role");
  if (roleAttribute) {
    const role = getStringLiteralAttributeValue(roleAttribute)?.toLowerCase();
    if (!role || REPEATED_TEXT_SKIPPED_ROLES.has(role)) return true;
  }
  return !isRoot && isTailwindCardSurface(openingElement);
};

const getStructuralPathSegment = (
  openingElement: EsTreeNodeOfType<"JSXOpeningElement">,
): string | null => {
  const elementName = getNativeElementName(openingElement);
  if (!elementName) return null;
  const classNameValue = getStringFromClassNameAttr(openingElement);
  if (!classNameValue) return elementName;
  const normalizedClasses = splitTailwindClassName(classNameValue).sort().join(".");
  return normalizedClasses ? `${elementName}.${normalizedClasses}` : elementName;
};

const normalizeRepeatedText = (value: string): string | null => {
  const text = value.replace(/\s+/g, " ").trim();
  if (
    text.length < REPEATED_CONTAINER_TEXT_MIN_LENGTH ||
    text.length > REPEATED_CONTAINER_TEXT_MAX_LENGTH ||
    !/\p{L}/u.test(text)
  ) {
    return null;
  }
  return text;
};

const appendTextOccurrence = (
  collection: RepeatedTextCollection,
  value: string,
  node: EsTreeNode,
  structuralPath: string[],
): void => {
  const text = normalizeRepeatedText(value);
  if (!text) return;
  const occurrences = collection.occurrencesByText.get(text) ?? [];
  occurrences.push({ node, signature: structuralPath.join(">") });
  collection.occurrencesByText.set(text, occurrences);
};

const collectStaticChild = (
  node: EsTreeNode,
  structuralPath: string[],
  collection: RepeatedTextCollection,
  context: RuleContext,
): void => {
  if (!collection.isStatic) return;
  if (isNodeOfType(node, "JSXText")) {
    appendTextOccurrence(collection, node.value ?? "", node, structuralPath);
    return;
  }
  if (isNodeOfType(node, "JSXElement")) {
    if (shouldSkipElement(node.openingElement, false, context)) return;
    collection.descendantCount += 1;
    if (collection.descendantCount > REPEATED_CONTAINER_TEXT_MAX_DESCENDANT_COUNT) {
      collection.isStatic = false;
      return;
    }
    if (hasResponsiveVisibility(node.openingElement)) {
      collection.isStatic = false;
      return;
    }
    const pathSegment = getStructuralPathSegment(node.openingElement);
    if (!pathSegment) return;
    const nextStructuralPath = [...structuralPath, pathSegment];
    for (const child of node.children) {
      collectStaticChild(child, nextStructuralPath, collection, context);
    }
    return;
  }
  if (isNodeOfType(node, "JSXFragment")) {
    for (const child of node.children) {
      collectStaticChild(child, structuralPath, collection, context);
    }
    return;
  }
  if (!isNodeOfType(node, "JSXExpressionContainer")) {
    collection.isStatic = false;
    return;
  }
  const expression = node.expression;
  if (isNodeOfType(expression, "Literal")) {
    if (typeof expression.value === "string") {
      appendTextOccurrence(collection, expression.value, expression, structuralPath);
    }
    return;
  }
  if (
    isNodeOfType(expression, "TemplateLiteral") &&
    expression.expressions.length === 0 &&
    expression.quasis.length === 1
  ) {
    appendTextOccurrence(
      collection,
      expression.quasis[0].value.cooked ?? expression.quasis[0].value.raw,
      expression,
      structuralPath,
    );
    return;
  }
  if (!isNodeOfType(expression, "JSXEmptyExpression")) collection.isStatic = false;
};

const collectRepeatedText = (
  container: EsTreeNodeOfType<"JSXElement">,
  context: RuleContext,
): RepeatedTextCollection | null => {
  if (
    hasResponsiveVisibility(container.openingElement) ||
    shouldSkipElement(container.openingElement, true, context)
  ) {
    return null;
  }
  const collection: RepeatedTextCollection = {
    descendantCount: 0,
    isStatic: true,
    occurrencesByText: new Map(),
  };
  for (const child of container.children) collectStaticChild(child, [], collection, context);
  return collection.isStatic ? collection : null;
};

export const noRepeatedContainerText = defineRule({
  id: "no-repeated-container-text",
  title: "A card repeats the same text in distinct slots",
  severity: "warn",
  defaultEnabled: false,
  tags: ["design", "test-noise", "react-jsx-only"],
  category: "Design",
  recommendation:
    "Keep repeated status or label copy in the one card slot where it carries the clearest meaning.",
  create: (context: RuleContext) => ({
    JSXElement(node: EsTreeNodeOfType<"JSXElement">) {
      if (
        !hasCapabilityOrUnspecified(context.settings, "tailwind") ||
        !isNodeOfType(node.openingElement.name, "JSXIdentifier") ||
        !REPEATED_TEXT_CONTAINER_NAMES.has(node.openingElement.name.name) ||
        !isTailwindCardSurface(node.openingElement)
      ) {
        return;
      }
      const collection = collectRepeatedText(node, context);
      if (!collection) return;
      for (const [text, occurrences] of collection.occurrencesByText) {
        const distinctSignatures = new Set(occurrences.map((occurrence) => occurrence.signature));
        if (
          occurrences.length < REPEATED_CONTAINER_TEXT_MIN_COUNT ||
          distinctSignatures.size < REPEATED_CONTAINER_TEXT_MIN_COUNT
        ) {
          continue;
        }
        context.report({
          node: occurrences[0].node,
          message: `The literal "${text}" appears in ${distinctSignatures.size} structurally different spots inside this card. Keep it in the one slot where it matters most.`,
        });
      }
    },
  }),
});

File: native/oxlint/rules/no-repeated-container-text.rs
use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};
use oxc_ast::{
    AstKind,
    ast::{Expression, JSXAttributeItem, JSXChild},
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

#[derive(Debug, Default, Clone)]
pub struct NoRepeatedContainerText;
declare_oxc_lint!(
    /// Disallow repeated card text in distinct structural slots.
    NoRepeatedContainerText,
    react_doctor_native,
    style,
    version = "0.1.0",
    short_description = "Disallow repeated card text in distinct structural slots."
);

#[derive(Default)]
struct RepeatedTextCollection {
    descendants: usize,
    is_static: bool,
    occurrences: rustc_hash::FxHashMap<String, Vec<(oxc_span::Span, String)>>,
}

impl Rule for NoRepeatedContainerText {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx() && !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::JSXElement(element) = node.kind() else {
            return;
        };
        let Some(name) = no_repeated_name(&element.opening_element) else {
            return;
        };
        if !matches!(name, "article" | "aside" | "div" | "section")
            || !has_capability_or_unspecified(ctx, "tailwind")
            || !is_tailwind_card_surface(&element.opening_element)
            || no_repeated_responsive(&element.opening_element)
            || no_repeated_skip(&element.opening_element, true, ctx)
        {
            return;
        }
        let mut collection = RepeatedTextCollection {
            is_static: true,
            ..Default::default()
        };
        for child in &element.children {
            no_repeated_collect(child, &mut Vec::new(), &mut collection, ctx);
        }
        if !collection.is_static {
            return;
        }
        for (text, occurrences) in collection.occurrences {
            let signatures = occurrences
                .iter()
                .map(|(_, signature)| signature)
                .collect::<rustc_hash::FxHashSet<_>>();
            if occurrences.len() >= 3 && signatures.len() >= 3 {
                ctx.diagnostic(OxcDiagnostic::warn(format!("The literal \"{text}\" appears in {} structurally different spots inside this card. Keep it in the one slot where it matters most.", signatures.len())).with_label(occurrences[0].0));
            }
        }
    }
}
fn no_repeated_name<'a>(opening: &'a oxc_ast::ast::JSXOpeningElement<'a>) -> Option<&'a str> {
    match &opening.name {
        oxc_ast::ast::JSXElementName::Identifier(identifier)
            if identifier.name == identifier.name.to_ascii_lowercase() =>
        {
            Some(identifier.name.as_str())
        }
        _ => None,
    }
}
fn no_repeated_responsive(opening: &oxc_ast::ast::JSXOpeningElement<'_>) -> bool {
    let Some(class_name) = get_static_class_name(opening) else {
        return false;
    };
    tailwind_class_name_tokens(class_name).iter().any(|token| {
        let visible = !matches!(
            get_tailwind_visibility_effect(token.utility),
            TailwindVisibilityEffectResolution::NotRelevant
        ) || matches!(token.utility, "sr-only" | "not-sr-only");
        visible
            && token.variants.iter().any(|variant| {
                matches!(*variant, "sm" | "md" | "lg" | "xl" | "2xl")
                    || variant.strip_prefix("max-").is_some_and(|breakpoint| {
                        matches!(breakpoint, "sm" | "md" | "lg" | "xl" | "2xl")
                    })
                    || variant
                        .strip_prefix("min-[")
                        .and_then(|value| value.strip_suffix(']'))
                        .is_some_and(|value| !value.is_empty())
                    || variant
                        .strip_prefix("max-[")
                        .and_then(|value| value.strip_suffix(']'))
                        .is_some_and(|value| !value.is_empty())
            })
    })
}
fn no_repeated_hidden<'a>(
    opening: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    if let Some(attribute) = get_authoritative_jsx_attribute(opening, "hidden", true) {
        if attribute.value.is_none()
            || get_string_literal_attribute_value(attribute)
                .is_none_or(|value| !value.eq_ignore_ascii_case("false"))
        {
            return true;
        }
    }
    if let Some(attribute) = get_authoritative_jsx_attribute(opening, "aria-hidden", true) {
        if get_string_literal_attribute_value(attribute)
            .is_none_or(|value| value.eq_ignore_ascii_case("true"))
        {
            return true;
        }
    }
    if let Some(style_attribute) = get_authoritative_jsx_attribute(opening, "style", true) {
        let Some(style) = get_inline_style_object_expression_with_aliases(style_attribute, ctx)
        else {
            return false;
        };
        for name in ["display", "visibility"] {
            if let Some(property) = get_effective_static_style_property(style, name) {
                let value = get_object_property_string_value(property);
                if value.is_none()
                    || value.is_some_and(|value| {
                        matches!(
                            value.trim().to_ascii_lowercase().as_str(),
                            "none" | "hidden" | "collapse"
                        )
                    })
                {
                    return true;
                }
            }
        }
    }
    let Some(class_name) = get_static_class_name(opening) else {
        return false;
    };
    let tokens = tailwind_class_name_tokens(class_name);
    if tokens.iter().any(|token| {
        matches!(
            token.utility,
            "screen-reader-only" | "sr-only" | "visually-hidden"
        )
    }) || get_effective_tailwind_class_name_token(&tokens, |utility| {
        matches!(utility, "sr-only" | "not-sr-only")
    }) == Some("sr-only")
    {
        return true;
    }
    get_tailwind_visibility_at_breakpoints(class_name)
        .is_some_and(|values| values.iter().all(|value| !value))
}
fn no_repeated_skip<'a>(
    opening: &'a oxc_ast::ast::JSXOpeningElement<'a>,
    root: bool,
    ctx: &LintContext<'a>,
) -> bool {
    let Some(name) = no_repeated_name(opening) else {
        return true;
    };
    if matches!(
        name,
        "a" | "button"
            | "canvas"
            | "code"
            | "datalist"
            | "dd"
            | "dl"
            | "dt"
            | "figure"
            | "input"
            | "kbd"
            | "label"
            | "menu"
            | "nav"
            | "ol"
            | "option"
            | "optgroup"
            | "pre"
            | "samp"
            | "select"
            | "summary"
            | "svg"
            | "table"
            | "tbody"
            | "td"
            | "textarea"
            | "tfoot"
            | "th"
            | "thead"
            | "tr"
            | "ul"
    ) {
        return true;
    }
    if opening.attributes.iter().any(|attribute| matches!(attribute, JSXAttributeItem::SpreadAttribute(_)) || matches!(attribute, JSXAttributeItem::Attribute(attribute) if matches!(&attribute.name, oxc_ast::ast::JSXAttributeName::Identifier(identifier) if matches!(identifier.name.as_str(), "children" | "dangerouslySetInnerHTML")))) { return true; }
    let class_attribute = get_authoritative_jsx_attribute(opening, "className", true);
    let class_name = get_static_class_name(opening);
    if class_attribute.is_some() && class_name.is_none()
        || class_name.is_some_and(no_repeated_data_visualization_class)
        || no_repeated_hidden(opening, ctx)
    {
        return true;
    }
    if let Some(role) = get_authoritative_jsx_attribute(opening, "role", true) {
        let Some(role) = get_string_literal_attribute_value(role) else {
            return true;
        };
        if role.is_empty()
            || matches!(
                role.to_ascii_lowercase().as_str(),
                "button"
                    | "cell"
                    | "grid"
                    | "gridcell"
                    | "graphics-document"
                    | "graphics-symbol"
                    | "img"
                    | "link"
                    | "list"
                    | "listbox"
                    | "listitem"
                    | "menu"
                    | "menubar"
                    | "navigation"
                    | "progressbar"
                    | "radiogroup"
                    | "row"
                    | "rowgroup"
                    | "table"
                    | "tablist"
                    | "tree"
                    | "treeitem"
                    | "diagram"
            )
        {
            return true;
        }
    }
    !root && is_tailwind_card_surface(opening)
}
fn no_repeated_data_visualization_class(value: &str) -> bool {
    value
        .to_ascii_lowercase()
        .split(|character: char| character == '-' || character == '_' || character.is_whitespace())
        .any(|part| {
            matches!(
                part,
                "chart" | "graph" | "heatmap" | "plot" | "visualization"
            )
        })
}
fn no_repeated_signature(opening: &oxc_ast::ast::JSXOpeningElement<'_>) -> Option<String> {
    let name = no_repeated_name(opening)?;
    let Some(class_name) = get_static_class_name(opening) else {
        return Some(name.to_string());
    };
    let parsed_tokens = tailwind_class_name_tokens(class_name);
    if parsed_tokens.is_empty() {
        return Some(name.to_string());
    }
    let mut tokens = parsed_tokens
        .iter()
        .map(|token| token.raw_token)
        .collect::<Vec<_>>();
    tokens.sort_unstable();
    Some(format!("{name}.{}", tokens.join(".")))
}
fn no_repeated_append(
    value: &str,
    span: oxc_span::Span,
    path: &[String],
    collection: &mut RepeatedTextCollection,
) {
    let text = value.split_whitespace().collect::<Vec<_>>().join(" ");
    let text_length = text.encode_utf16().count();
    if text_length < 4 || text_length > 48 || !text.chars().any(char::is_alphabetic) {
        return;
    }
    collection
        .occurrences
        .entry(text)
        .or_default()
        .push((span, path.join(">")));
}
fn no_repeated_collect<'a>(
    child: &'a JSXChild<'a>,
    path: &mut Vec<String>,
    collection: &mut RepeatedTextCollection,
    ctx: &LintContext<'a>,
) {
    if !collection.is_static {
        return;
    }
    match child {
        JSXChild::Text(text) => {
            no_repeated_append(text.value.as_str(), text.span, path, collection)
        }
        JSXChild::Element(element) => {
            if no_repeated_skip(&element.opening_element, false, ctx) {
                return;
            }
            collection.descendants += 1;
            if collection.descendants > 250 || no_repeated_responsive(&element.opening_element) {
                collection.is_static = false;
                return;
            }
            let Some(segment) = no_repeated_signature(&element.opening_element) else {
                return;
            };
            path.push(segment);
            for child in &element.children {
                no_repeated_collect(child, path, collection, ctx);
            }
            path.pop();
        }
        JSXChild::Fragment(fragment) => {
            for child in &fragment.children {
                no_repeated_collect(child, path, collection, ctx);
            }
        }
        JSXChild::ExpressionContainer(container) => match container
            .expression
            .as_expression()
            .map(Expression::get_inner_expression)
        {
            Some(Expression::StringLiteral(value)) => {
                no_repeated_append(value.value.as_str(), value.span, path, collection)
            }
            Some(Expression::TemplateLiteral(template)) if template.expressions.is_empty() => {
                if let Some(quasi) = template.quasis.first() {
                    no_repeated_append(
                        quasi
                            .value
                            .cooked
                            .as_ref()
                            .map_or(quasi.value.raw.as_str(), |value| value.as_str()),
                        template.span,
                        path,
                        collection,
                    );
                }
            }
            Some(
                Expression::BooleanLiteral(_)
                | Expression::NullLiteral(_)
                | Expression::NumericLiteral(_)
                | Expression::BigIntLiteral(_)
                | Expression::RegExpLiteral(_),
            ) => {}
            None => {}
            _ => collection.is_static = false,
        },
        _ => collection.is_static = false,
    }
}
